# 📋 Relatório 01 — A Régua do Observatório

**Bloco 1 · Aulas 1–2 · Unidade 1 — Fundamentos de Lógica de Programação**
**Prazo:** domingo, **30/08/2026**, até 23h59 · **Vale 10,0** (compõe os 30%
de *Exercícios práticos por bloco*)

---

Nas Missões 1 e 2 você construiu as **calculadoras** e a **Mesa de Decisões**
do Observatório Econômico Municipal — juros compostos, equação de Fisher,
CAGR, IRPF progressivo, semáforos de risco — sempre sobre os mesmos números,
os mesmos para a turma inteira.

Aqui o trabalho é outro: **cada analista recebe um município para
acompanhar**, e o relatório aplica a ele todas as réguas do observatório — a
do desenvolvimento humano, a do tempo, a do poder de compra, a da meta, a
fiscal.

O seu município, os seus indicadores, a sua série histórica e o seu ano foco
são derivados da sua **matrícula**. Ninguém da turma tem os mesmos números que
você — e é por isso que o notebook do colega não serve, mesmo que o código
dele esteja perfeito.

### 🤖 Política de IA nesta atividade

**Unidade 1 — uso de IA generativa não permitido** (§5.2 do programa). Nesta
unidade a lógica de programação é construída "à mão": é ela que sustenta tudo
o que vem depois. A declaração de não-uso, ao final, é obrigatória.

Você pode e deve consultar: os módulos de slides, os notebooks das Missões, a
documentação oficial do Python e os colegas — desde que o código que você
entrega seja seu e você saiba explicá-lo.

### 📤 Como entregar — `git push` + link no Google Sala de Aula

Este repositório é **seu**: você criou clicando em **"Use this template"** na
página do repositório da disciplina no GitHub. A entrega tem duas partes — o
`git push` guarda o seu trabalho, mas **só conta como entregue depois que o
link do repositório for enviado no Google Sala de Aula**.

1. Preencha **todas** as células marcadas com `# TODO` e todos os campos de
   texto (previsão, questão autoral, relato do obstáculo, reflexão,
   declaração).
2. Rode o notebook **inteiro, de cima para baixo**, e **salve**. As saídas
   precisam estar visíveis no arquivo — é isso que o professor lê.
3. Faça o commit final com a linha `ASSINATURA:` como mensagem, e dê push
   para a branch `main` do **seu** repositório:

```bash
git add atividade-01-a-regua-do-observatorio.ipynb
git commit -m "ASSINATURA: <cole aqui a linha da última célula>"
git push
```

4. Vá ao **Google Sala de Aula**, abra a atividade "Relatório 01" e cole o
   endereço do seu repositório no GitHub. **É esse link que registra a
   entrega** — sem ele, o professor não sabe que você terminou, mesmo que o
   push tenha funcionado.

> 💡 **Commits durante o trabalho são bem-vindos** — commit a cada exercício
> resolvido, se quiser. O histórico do repositório é o registro mais honesto
> do seu processo, e nada nele é penalizado: um `git log` com idas e vindas é
> exatamente o que se espera de quem está aprendendo.
>
> ⏰ Vale o **último commit antes do prazo**. Você só precisa colar o link uma
> vez — se enviar antes de terminar, continue commitando e dando push até o
> prazo: é o último commit que entra na correção, não a hora em que o link
> foi colado.

### 🧭 Como esta atividade é avaliada

| Critério | Pontos |
|---|---|
| Correção técnica dos exercícios | 4,0 |
| Previsão e rastreamento (Parte 1) | 1,5 |
| Questão autoral | 2,0 |
| Diário de bordo + relato do obstáculo | 1,5 |
| Reflexão obrigatória + declaração de IA | 1,0 |

> 💡 Repare que o código vale **menos da metade**. Os outros 6,0 pontos estão
> no seu **raciocínio**: o que você previu antes de rodar, o obstáculo que
> você realmente enfrentou, a questão que você inventou. Esses pontos ninguém
> entrega por você.

> ⚠️ **Ferramentas desta atividade:** variáveis, os quatro tipos primitivos,
> operadores aritméticos, de comparação e lógicos, `print()` com f-strings,
> `if`/`elif`/`else`, `try`/`except` e `raise`. **Ainda não há listas, laços
> nem funções** — se você sentir falta delas, ótimo: é exatamente a falta que
> as próximas aulas vão preencher.
>
> Nas Missões, os dados entravam pelo `input()`. Aqui não: um notebook de
> entrega precisa rodar **inteiro, sem parar para digitar**. Quem faz o papel
> do teclado é o kit — e a conversão de texto para número, que era o ponto
> do `input()`, continua viva no Exercício 7.

---

## ⚙️ Passo 0 — Identificação

O kit deste relatório mora em `scripts/kit_r01.py`, ao lado deste notebook.
Ele deriva os seus dados da sua matrícula, mantém o diário de bordo e produz a
assinatura de entrega.

Você só precisa de **uma chamada**: `iniciar()`, passando a sua matrícula e o
seu nome. Depois disso, as variáveis do seu município simplesmente existem —
`MEU_MUNICIPIO`, `MEU_PIB_PC`, `MINHA_TAXA_NOMINAL` e as demais.

A matrícula é a semente de tudo: ela passa por um **resumo criptográfico
SHA-256**, e trocar um único dígito produz outro município e outra série
histórica inteira.

> 📖 O kit é legível de propósito. Se quiser saber de onde saem os seus
> números, abra `scripts/kit_r01.py` e leia — é Python comum, sem mágica.

In [2]:
from scripts.kit_r01 import (
    iniciar, prever, registrar, conferir, diario, assinatura
)

# TODO: troque pelos SEUS dados e rode a célula
iniciar(matricula="20261002340", nome="Rita_de_Cassia_Pereira_da_Costa")

📋 Relatório 01 — A Régua do Observatório  ·  kit r01 v2.0
Analista: Rita_de_Cassia_Pereira_da_Costa (20261002340)

🏙️  Município foco: Ipojuca (PE)
    MINHA_POPULACAO ..... 100.5 mil hab.
    MEU_IDH ............. 0.652
    MINHA_POBREZA ....... 34.9%

📈  Série do PIB per capita
    MEU_ANO_BASE ........ 2015
    MEU_PIB_PC_BASE ..... R$ 87745.54
    MEU_ANO ............. 2020
    MEU_PIB_PC .......... R$ 96597.25

💹  Mercado e plano diretor
    MINHA_TAXA_NOMINAL .. 6.16% a.a. (aplicação do superávit de caixa)
    MINHA_INFLACAO ...... 10.17% a.a.
    MEU_CRESCIMENTO ..... 0.88% a.a. (projeção do PIB per capita)
    MINHA_META .......... R$ 105600.00 (PIB per capita em 2030)

🏛️  Financiamento e dívida
    MINHAS_PARCELAS ..... 211 parcelas mensais
    MINHA_PARCELA ....... R$ 469.39 mil por mês
    MINHA_DIVIDA ........ '18.6'  <- repare no tipo: é TEXTO

✅ Kit ligado. As variáveis acima já existem — use-as direto.
📓 registrado: passo-0


Se a célula acima reclamou da matrícula, é porque o `"..."` continua lá — ela
precisa da sua matrícula completa do SIGAA, só dígitos, entre aspas.

As seis funções do kit, e é só isso que você vai usar dele:

| Função | Para quê |
|---|---|
| `iniciar(matricula, nome)` | liga o kit e cria os seus dados *(já feito)* |
| `prever(...)` | carimba a sua previsão **antes** de você revelar |
| `registrar(etapa, nota)` | marca uma etapa no seu diário de bordo |
| `conferir(etapa, ...)` | devolve um retorno sobre o que você resolveu |
| `diario()` | mostra o seu ritmo de trabalho |
| `assinatura()` | emite a linha de entrega |

> 💬 **Sobre o `conferir()`**: ele não quebra o notebook nem tira ponto. Olha o
> que você respondeu e diz, item a item, o que ainda não fecha — dando uma
> **pista**, nunca a resposta. Pode rodar quantas vezes quiser: errar ali não
> custa nada, é para isso que ele existe.

---

## 🔮 Parte 1 — Previsão (vale 1,5 — pelo raciocínio, não pelo acerto)

Antes de rodar qualquer coisa, **leia** o código abaixo e diga o que ele vai
produzir. Esta parte é avaliada pela **qualidade da justificativa**: uma
previsão errada, bem explicada, vale nota cheia. Uma previsão certa sem
explicação, não.

É o mesmo formato da Parte 1 (papel) do Checkpoint — então isto aqui também é
treino.

```python
projecao_a = MEU_PIB_PC * (1 + MEU_CRESCIMENTO / 100) ** 5
projecao_b = MEU_PIB_PC * 1 + MEU_CRESCIMENTO / 100 ** 5

print(f"A: {projecao_a:.2f}")
print(f"B: {projecao_b:.2f}")
```

As duas linhas usam exatamente os mesmos números e os mesmos operadores. Só
mudam os parênteses. É o bug do Bônus 1 da Missão 1 — agora sobre os **seus**
dados.

**Responda, sem rodar:**

1. Qual das duas linhas projeta corretamente o PIB per capita do seu município
   daqui a 5 anos?
2. Quanto vale, **aproximadamente**, `projecao_b`? (Não precisa da conta
   exata — precisa da ordem de grandeza e do motivo.)

Preencha a célula abaixo **primeiro**. Ela não corrige nada: só carimba, com a
hora, o que você achava que ia acontecer.

Respostas: 
1. A projeção correta é realizado peça linha 1, ou seja, pela `projecao_a`. Pois o **Python** realiza os calculos matemáticos seguindo a lógica, ou seja, inicialmente é realizado o calculo envolvendo exponencialção, seguidos por divisão, multiplicação, subtração/soma. Ao se acrescentar parênteses, como a projecao_a, o python entende que as preferencias de ordenamento mudaram, logo ele realizara os calculos dentro do parênteses e após isso, ele seguirar a sequencia mencionada anteriormente. 

2. Aproximadamente o resultado será **96597.25**, pois o **Python** realizará inicialmente o calculo exponencial, seguido pela divisão, $0.88/100 ^ 5$ resultando em um valor 0.0000, aredondando em 4 casas decimais, posteriormente será realizado a multiplicação $96597.25*1$, que não mudará o valor pois está multiplicando com 1. E por fim realizará a soma entre $ 96597.25 + 0.0000$ que resultará em **96597.2** 

$$ projecao_b = 96597.25*1 + 0.88/100 ^ 5 $$



In [3]:
# TODO: preencha ANTES de rodar a célula que revela
prever(
    linha_correta="projecao_a",        # "A" ou "B", entre aspas
    valor_aproximado=  100923    # um número
)

📓 registrado: previsao
🔮 Previsão carimbada. Agora sim: rode a célula seguinte.


### ✍️ Sua justificativa (é ela que vale os 1,5)

**Por quê — explique a ordem em que o Python avalia os operadores no chunk a seguir:**

> Passo 1 (`**`): O operador de exponenciação é o primeiro a ser avaliado pelo Python, a menos que, ao longo da expressão exista alguma inicação de ordem de prioridade dos cálculos, como por exemplo, os parênteses ( ), que possuem prioridades frente ao expoente.
>
> Passo 2 (`*` e `/`): Após a realização do calculo da exponenciação é realizado como prioridade a divisão seguida da multiplicação. Vale resaltar se houver a presença de parênteses, e dentro dele não houver expoentes, a divisão e a multiplicação ganham prioridade frente a soma e subtração.
>
> Passo 3 (`+`): Por fim, a soma, esse passo será o ultimo a ser realizado, se estiver dentro do parênteses, ele segue sendo o ultimo a ser realizado frente a expressão dentro do parênteses, mas ele cria prioridada para os calculos fora dele.
>
> Conclusão: O Python segue a ordem realizada dos calculos manualmente, logo é necessário indicar as prioridades dos calculos, e deixar sinalizado a ordem que ele deve realizar os calculos. 

In [4]:
# ▶️ Agora sim — rode e compare com a sua previsão
projecao_a = 96597.25 * (1 + 0.88 / 100) ** 5
projecao_b = 96597.25 * 1 + 0.88 / 100 ** 5

print(f"PIB per capita hoje: R$ {96597.25:.2f}")
print(f"A: R$ {projecao_a:.2f}")
print(f"B: R$ {projecao_b:.2f}")
print(f"Diferença entre as duas: R$ {projecao_a - projecao_b:.2f}")

# TODO: substitua a nota abaixo — acertei? o que me surpreendeu?
registrar("previsao-revelada", "100923")

PIB per capita hoje: R$ 96597.25
A: R$ 100923.00
B: R$ 96597.25
Diferença entre as duas: R$ 4325.75
📓 registrado: previsao-revelada


### 🔍 Depois de rodar

**A sua previsão se confirmou?** `Sim` *(sim / em parte / não)*

**O que você aprendeu com a diferença entre o que previu e o que saiu:**

> ...

> 💡 Guarde esta lição: em fórmulas financeiras, um parêntese esquecido **não
> gera erro** — gera um número errado que parece razoável. É o tipo de bug
> mais perigoso em análise de dados, porque nada avisa.

---

## 🥉 Exercício 1 — A ficha do município

Todo relatório do OEM abre com a ficha de identificação. Monte a sua usando
**os quatro tipos primitivos** e uma f-string, como no Aquecimento 2 da
Missão 1 — só que agora com os dados do seu município.

Duas contas entram aqui:

- **PIB total** = PIB per capita × população. Atenção às unidades:
  `MEU_PIB_PC` está em R$ por habitante e `MINHA_POPULACAO` em **mil**
  habitantes, então o produto sai em **mil R$**.
- **Município de grande porte?** O IBGE chama de grande porte o município com
  mais de 100 mil habitantes. Produza um `bool` com uma **comparação** — sem
  `if`, como no Aquecimento da Missão 1.

In [5]:
# TODO 1: o PIB total do seu município, em mil R$
pib_total = (MEU_PIB_PC * MINHA_POPULACAO) 

# TODO 2: um bool a partir de uma comparação (sem if!)
#         MINHA_POPULACAO está em MIL habitantes — cuidado com o corte
eh_grande_porte = MINHA_POPULACAO >= 100

# TODO 3: monte a ficha em UMA f-string, seguindo este modelo:
# "João Pessoa (PB) | 833.0 mil hab. | PIB per capita: R$ 27800.00 | IDH 0.76 | grande porte: True"
ficha = f"{MEU_MUNICIPIO} (PE)| {MINHA_POPULACAO:.1f} mil hab. | PIB per capita: R$ {MEU_PIB_PC:.2f} | IDH {MEU_IDH} | grande porte: {eh_grande_porte}"

# TODO 4: confirme os tipos das variáveis do kit
print(type(MEU_MUNICIPIO), type(MINHA_POPULACAO), type(MEU_ANO), type(eh_grande_porte))

print(ficha)
print(f"PIB total: R$ {pib_total:.2f} mil")

<class 'str'> <class 'float'> <class 'int'> <class 'bool'>
Ipojuca (PE)| 100.5 mil hab. | PIB per capita: R$ 96597.25 | IDH 0.652 | grande porte: True
PIB total: R$ 9708023.62 mil


In [6]:
conferir(
    "exercicio-1",
    pib_total=pib_total,
    eh_grande_porte=eh_grande_porte,
    ficha=ficha,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício —
# o que travou, o que você tentou, o que entendeu
registrar("exercicio-1", "É necessário ter atenção as unidades de medidas, pois cada variável está em uma unidade diferente, e ao realizar a combinação entre elas, é preciso ter atenção para entrender qual unidade de medida ressultará da operação. ")

🔎 CONFERÊNCIA — exercicio-1
   ✅ pib_total
   ✅ eh_grande_porte
   ✅ ficha
   --------------------------------------------------------------
   🎉 3/3 — etapa fechada.
📓 registrado: exercicio-1


---

## 🥉 Exercício 2 — A régua do desenvolvimento humano

A régua do PNUD classifica o desenvolvimento humano em quatro faixas:

| Faixa | Critério |
|---|---|
| muito alto | IDH ≥ 0,800 |
| alto | 0,700 ≤ IDH < 0,800 |
| médio | 0,550 ≤ IDH < 0,700 |
| baixo | IDH < 0,550 |

Escreva a cadeia `if`/`elif`/`else` que atribui a faixa correta ao **seu**
município.

Duas exigências, e as duas valem ponto:

1. testando **de cima para baixo**, você não precisa escrever a condição dupla
   (`>= 0.700 and < 0.800`) — se o `if` anterior falhou, o Python já sabe que
   o IDH é menor que 0,800;
2. a cadeia precisa terminar em `else`, sem nenhuma condição. Um `if`/`elif`
   sem `else` deixa um buraco: existe entrada para a qual nada é atribuído, e
   a linha seguinte quebra com `NameError`. **Isso vale mesmo que o seu
   município nunca caia no último ramo** — o `else` protege o código, não o
   caso particular que você tem em mãos.

In [7]:
# TODO 1: complete a cadeia de decisão
if MEU_IDH >= 0.8:
    faixa_idh = "muito alto"
elif 0.7 <= MEU_IDH < 0.8:
    faixa_idh = "alto"
elif 0.550 <= MEU_IDH < 0.7:
    faixa_idh = "médio"
else:
    faixa_idh = "baixo"

print(f"{MEU_MUNICIPIO}: IDH {MEU_IDH} -> desenvolvimento humano {faixa_idh}")

# TODO 2: quantos pontos faltam para o seu município chegar à faixa
# "muito alto"? Se ele já estiver lá, o resultado deve ser 0 (nunca negativo).
# Use um if/else — e pense qual dos dois lados precisa da conta.
if MEU_IDH >= 0.8:
    falta_para_muito_alto = 0.0
else:
    falta_para_muito_alto = 0.8 - MEU_IDH

print(f"Faltam {falta_para_muito_alto:.3f} pontos para 'muito alto'.")

Ipojuca: IDH 0.652 -> desenvolvimento humano médio
Faltam 0.148 pontos para 'muito alto'.


In [ ]:
conferir(
    "exercicio-2",
    faixa_idh=faixa_idh,
    falta_para_muito_alto=falta_para_muito_alto,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício
registrar("exercicio-2", "A lógica a ser seguida para a construção do mais específico para o geral, foi necessário atenção, principalmente na segunda parte, para a condição do else")

🔎 CONFERÊNCIA — exercicio-2
   ✅ faixa_idh
   ✅ falta_para_muito_alto
   --------------------------------------------------------------
   🎉 2/2 — etapa fechada.
📓 registrado: exercicio-2


---

## 🥈 Exercício 3 — A régua do tempo: quanto o seu município cresceu de fato

O kit te deu **dois pontos** da série do PIB per capita: `MEU_PIB_PC_BASE` no
ano `MEU_ANO_BASE`, e `MEU_PIB_PC` no ano `MEU_ANO`. A pergunta do relatório é
a do Desafio 1 da Missão 1: **qual foi o crescimento médio por ano?**

E a resposta **não** é "crescimento total ÷ número de anos" — porque o
crescimento se capitaliza. A taxa média geométrica (*CAGR*) é:

$$\text{CAGR} = \left(\frac{V_{final}}{V_{inicial}}\right)^{1/n} - 1$$

> ⚠️ Conte os **intervalos**, não os anos: de 2019 a 2023 são **4** períodos
> de crescimento, não 5. É o erro clássico, e ele contamina o CAGR inteiro.

> 💡 O seu crescimento pode ter sido **negativo** — há município brasileiro
> que encolheu no período, e a fórmula lida com isso naturalmente. Se der
> negativo, não é bug: é economia.

No fim, a comparação que interessa ao gestor: o plano diretor projeta
`MEU_CRESCIMENTO`% ao ano daqui para a frente. Isso é mais otimista do que o
município de fato entregou?

In [ ]:
# TODO 1: quantos ANOS de crescimento separam os dois pontos da série
anos_da_serie = ...

# TODO 2: o crescimento total do período, em %
crescimento_total_pct = ...

# TODO 3: o CAGR, em % ao ano (dica: expoente fracionário com **)
cagr_pct = ...

# TODO 4: a média aritmética "ingênua" que muita gente reporta por engano
media_ingenua_pct = ...

# TODO 5: um bool, sem if — a projeção do plano diretor (MEU_CRESCIMENTO)
# é mais otimista que o CAGR que o município realmente entregou?
projecao_otimista = ...

print(f"{MEU_MUNICIPIO}, de {MEU_ANO_BASE} a {MEU_ANO} ({anos_da_serie} anos):")
print(f"  PIB per capita ....... R$ {MEU_PIB_PC_BASE:.2f} -> R$ {MEU_PIB_PC:.2f}")
print(f"  crescimento total .... {crescimento_total_pct:.2f}%")
print(f"  CAGR (correto) ....... {cagr_pct:.2f}% ao ano")
print(f"  média ingênua (errada) {media_ingenua_pct:.2f}% ao ano")
print(f"  projeção do plano diretor: {MEU_CRESCIMENTO}% ao ano")
print("  Plano diretor mais otimista que o histórico?", projecao_otimista)

In [ ]:
conferir(
    "exercicio-3",
    anos_da_serie=anos_da_serie,
    crescimento_total_pct=crescimento_total_pct,
    cagr_pct=cagr_pct,
    media_ingenua_pct=media_ingenua_pct,
    projecao_otimista=projecao_otimista,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício
registrar("exercicio-3", "...")

---

## 🥈 Exercício 4 — A régua do poder de compra: a equação de Fisher

A prefeitura aplicou o superávit de caixa e rendeu `MINHA_TAXA_NOMINAL`% no
ano. No mesmo período, a inflação foi de `MINHA_INFLACAO`%. **O município
ficou mais rico?**

O ganho **real** não é a subtração simples: a fórmula correta é a **equação de
Fisher exata**, a mesma da Prática 2 da Missão 1.

$$1 + r = \frac{1 + i}{1 + \pi} \qquad\Longrightarrow\qquad
r = \frac{1 + i}{1 + \pi} - 1$$

onde $r$ = taxa real, $i$ = taxa nominal, $\pi$ = inflação — todas em
**fração**, não em percentual. As suas duas variáveis chegam em percentual;
converter é com você.

Calcule as duas versões — a exata e a aproximação de bolso ($i - \pi$) — e
mostre em quantos **pontos percentuais** elas discordam. Depois, o semáforo do
Aquecimento 1 da Missão 2:

| Situação | Sinal |
|---|---|
| $r > 2\%$ | 🟢 ganho real relevante |
| $0 < r \le 2\%$ | 🟡 ganho real modesto |
| $r \le 0$ | 🔴 perda de poder de compra |

In [ ]:
# TODO 1: Fisher exata, resultado em PERCENTUAL
#         (converta as duas taxas para fração antes; multiplique por 100 depois)
taxa_real_pct = ...

# TODO 2: a aproximação de bolso, em percentual
taxa_real_aproximada_pct = ...

# TODO 3: a diferença, em pontos percentuais — quanto a aproximação
#         SUPERESTIMA o ganho real. O sinal importa.
erro_pp = ...

# TODO 4: o semáforo, com if/elif/else
if ...:
    sinal_fisher = "🟢 ganho real relevante"
elif ...:
    sinal_fisher = "🟡 ganho real modesto"
else:
    sinal_fisher = "🔴 perda de poder de compra"

print(f"Nominal: {MINHA_TAXA_NOMINAL:.2f}%  |  Inflação: {MINHA_INFLACAO:.2f}%")
print(f"  taxa real (Fisher exata) .. {taxa_real_pct:.2f}%")
print(f"  taxa real (aproximação) ... {taxa_real_aproximada_pct:.2f}%")
print(f"  a aproximação superestima em {erro_pp:.2f} p.p.")
print(f"Parecer sobre a aplicação: {sinal_fisher}")

In [ ]:
conferir(
    "exercicio-4",
    taxa_real_pct=taxa_real_pct,
    taxa_real_aproximada_pct=taxa_real_aproximada_pct,
    erro_pp=erro_pp,
    sinal_fisher=sinal_fisher,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício
registrar("exercicio-4", "...")

---

## 🥈 Exercício 5 — A régua da meta: a projeção até 2030

O plano diretor do seu município fixou uma meta de PIB per capita para
**2030** — é a variável `MINHA_META`, em R$ por habitante. Partindo do seu ano
foco (`MEU_ANO`) e do crescimento projetado (`MEU_CRESCIMENTO`), a meta é
alcançável?

A projeção usa a mesma fórmula dos juros compostos do Aquecimento 1 da
Missão 1:

$$V_{2030} = V_{\text{ano foco}} \times \left(1 + \frac{g}{100}\right)^{h}$$

onde $h$ é o **horizonte**, em anos, entre o ano foco e 2030 — contado como no
Exercício 3.

Depois da projeção, uma cadeia de três faixas para o parecer:

| Situação | Parecer |
|---|---|
| projeção ≥ meta | ✅ meta alcançada |
| entre 90% da meta e a meta | ⚠️ meta próxima, exige aceleração |
| abaixo de 90% da meta | ❌ meta inalcançável no ritmo atual |

In [ ]:
# TODO 1: o horizonte, em anos, entre o seu ano foco e 2030
horizonte = ...

# TODO 2: a projeção (atenção aos parênteses — lembre da Parte 1!)
pib_projetado = ...

# TODO 3: quanto isso representa da meta, em %
percentual_da_meta = ...

# TODO 4: o parecer, com if/elif/else
if ...:
    parecer = "✅ meta alcançada"
elif ...:
    parecer = "⚠️ meta próxima, exige aceleração"
else:
    parecer = "❌ meta inalcançável no ritmo atual"

print(f"De {MEU_ANO} a 2030 são {horizonte} anos de crescimento.")
print(f"PIB per capita projetado: R$ {pib_projetado:.2f}")
print(f"Isso é {percentual_da_meta:.1f}% da meta de R$ {MINHA_META:.2f}")
print(f"Parecer para {MEU_MUNICIPIO}: {parecer}")

In [ ]:
conferir(
    "exercicio-5",
    horizonte=horizonte,
    pib_projetado=pib_projetado,
    percentual_da_meta=percentual_da_meta,
    parecer=parecer,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício
registrar("exercicio-5", "...")

---

## 🥇 Exercício 6 — A régua fiscal: alíquota marginal × alíquota efetiva

Cada município contribui para o **Fundo Regional de Desenvolvimento** com uma
alíquota que cresce por faixa de PIB per capita — exatamente o desenho da
tabela do IRPF que você viu na Prática 2 da Missão 2, com **parcela a
deduzir** e tudo:

| Faixa de PIB per capita | Alíquota | Parcela a deduzir |
|---|---|---|
| até R$ 18.000 | isento | — |
| acima de R$ 18.000 até R$ 28.000 | 1,0% | R$ 180,00 |
| acima de R$ 28.000 até R$ 45.000 | 2,5% | R$ 600,00 |
| acima de R$ 45.000 | 4,5% | R$ 1.500,00 |

$$\text{contribuição por habitante} = \text{PIB per capita} \times
\text{alíquota} - \text{parcela a deduzir}$$

E aqui mora a confusão que todo economista precisa saber desmontar em 30
segundos — **duas alíquotas convivem, e elas não são a mesma coisa**:

- a **marginal** é a da tabela, a que incide sobre o *próximo* real de PIB;
- a **efetiva** é o que o município de fato paga: contribuição ÷ PIB per
  capita.

> 💡 Repare que a alíquota e a parcela a deduzir **andam sempre juntas**: são
> duas atribuições no mesmo ramo do `if`. Trocar uma sem trocar a outra é o
> bug mais comum desta tabela.

In [ ]:
# TODO 1 e 2: a alíquota marginal E a parcela a deduzir, com if/elif/else.
# A cadeia vai da MENOR faixa para a maior, porque o operador é <=
if ...:
    aliquota = 0.000
    parcela_deduzir = 0.00
elif ...:
    aliquota = 0.010
    parcela_deduzir = 180.00
elif ...:
    aliquota = 0.025
    parcela_deduzir = 600.00
else:
    aliquota = 0.045
    parcela_deduzir = 1500.00

# TODO 3: a contribuição por habitante, em R$
contribuicao_por_habitante = ...

# TODO 4: a contribuição total do município, em mil R$
#         (população em mil hab. × R$ por hab. = mil R$)
contribuicao_total = ...

# TODO 5: a alíquota EFETIVA, em % — o que o município realmente paga
aliquota_efetiva_pct = ...

# TODO 6: um bool combinando DUAS condições com `and`, sem nenhum if.
# O município é contribuinte relevante se paga alíquota marginal acima de
# 1,0% E contribui com mais de R$ 50.000 mil ao fundo.
contribuinte_relevante = ...

print(f"{MEU_MUNICIPIO}: PIB per capita R$ {MEU_PIB_PC:.2f}")
print(f"  alíquota marginal ..... {aliquota * 100:.2f}%")
print(f"  parcela a deduzir ..... R$ {parcela_deduzir:.2f}")
print(f"  por habitante ......... R$ {contribuicao_por_habitante:.2f}")
print(f"  contribuição total .... R$ {contribuicao_total:.2f} mil")
print(f"  alíquota EFETIVA ...... {aliquota_efetiva_pct:.2f}%")
print("Contribuinte relevante?", contribuinte_relevante)

**A pergunta que fecha o exercício** — responda no texto, olhando os seus dois
números:

**Por que a alíquota efetiva do seu município é menor que a marginal** (ou
igual a ela, se você caiu na faixa isenta)? **O que aconteceria com a
contribuição de um município na fronteira de duas faixas se a coluna "parcela
a deduzir" simplesmente não existisse?**

> ...

In [ ]:
conferir(
    "exercicio-6",
    aliquota=aliquota,
    parcela_deduzir=parcela_deduzir,
    contribuicao_por_habitante=contribuicao_por_habitante,
    contribuicao_total=contribuicao_total,
    aliquota_efetiva_pct=aliquota_efetiva_pct,
    contribuinte_relevante=contribuinte_relevante,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício
registrar("exercicio-6", "...")

---

## 🥇 Exercício 7 — O semáforo fiscal blindado

Este é o exercício que junta tudo o que a Missão 2 apresentou.

A dívida consolidada do seu município chega do portal da transparência **como
texto** — a variável `MINHA_DIVIDA` é uma `str`, como você viu no Passo 0. E
dado público chega sujo: pode vir `"n/d"`, pode vir vazio, pode vir com um
valor economicamente impossível.

Sua rotina precisa sobreviver a **dois tipos de problema diferentes**, e cada
um pede uma ferramenta:

| Problema | Exemplo | Ferramenta |
|---|---|---|
| Dado ilegível | `"n/d"`, `""` | `try` / `except ValueError` |
| Regra de negócio violada | `"-5"`, `"450"` | `raise ValueError` |

A regra de negócio: uma dívida em percentual da receita fora do intervalo
**[0, 300]** é dado corrompido, não município endividado.

E o semáforo, para o dado que passar na triagem:

| Situação | Sinal |
|---|---|
| dívida ≤ 40% | 🟢 risco baixo |
| acima de 40% até 60% | 🟡 risco moderado |
| acima de 60% | 🔴 risco alto |

> 💡 O padrão é o da Missão 2: o `try`/`except` responde *"o dado existe e é
> legítimo?"*; a cadeia `if`/`elif`/`else`, **dentro do `try`**, responde *"em
> que faixa ele cai?"*. Um único `except ValueError` cobre os dois problemas,
> porque o seu `raise` produz o mesmo tipo de exceção que o `float()` gera.

In [ ]:
# O dado como ele chega do portal — texto, não número
divida_texto = MINHA_DIVIDA
divida_pct = None

try:
    # TODO 1: converta o texto para número
    divida_pct = ...

    # TODO 2: a regra de negócio — levante ValueError com mensagem clara
    if ...:
        raise ValueError(...)

    # TODO 3: o semáforo, com if/elif/else
    if ...:
        sinal = "🟢 risco baixo"
    elif ...:
        sinal = "🟡 risco moderado"
    else:
        sinal = "🔴 risco alto"

except ValueError as erro:
    # TODO 4: exiba a mensagem capturada e marque o caso como sem dado
    sinal = "⚪ sem dado"
    print(f"⚠️ Triagem: {erro}")

print(f"{MEU_MUNICIPIO} | dívida informada: {divida_texto!r} -> {sinal}")

**Teste a sua rotina.** Na célula abaixo, troque `texto_teste` pelos cinco
casos e anote o que aconteceu em cada um. Os dois últimos precisam cair no
`except` — mas por caminhos diferentes, e é justamente essa diferença que o
exercício quer que você perceba.

In [ ]:
# TODO: rode esta célula uma vez para cada valor, trocando `texto_teste`,
# e anote o resultado ao lado.
#
#   MINHA_DIVIDA  -> sinal: ______   (o seu dado real)
#   "25"          -> sinal: ______
#   "78"          -> sinal: ______
#   "n/d"         -> sinal: ______   quem reclamou: o Python ou o seu raise?
#   "-5"          -> sinal: ______   quem reclamou: o Python ou o seu raise?

texto_teste = "n/d"      # <- troque aqui

try:
    pct_teste = float(texto_teste)
    if pct_teste < 0 or pct_teste > 300:
        raise ValueError(f"dívida de {pct_teste}% da receita é implausível")
    if pct_teste <= 40:
        sinal_teste = "🟢 risco baixo"
    elif pct_teste <= 60:
        sinal_teste = "🟡 risco moderado"
    else:
        sinal_teste = "🔴 risco alto"
except ValueError as erro:
    sinal_teste = "⚪ sem dado"
    print(f"⚠️ Triagem: {erro}")

print(f"{texto_teste!r} -> {sinal_teste}")

# 👀 Repare que esta célula usa nomes próprios (`texto_teste`, `pct_teste`)
# para não sobrescrever o resultado oficial do exercício acima.

In [ ]:
conferir(
    "exercicio-7",
    divida_pct=divida_pct,
    sinal=sinal,
)

# TODO: substitua a nota abaixo por uma frase sua sobre este exercício
registrar("exercicio-7", "...")

---

## 🏆 Bônus — O prazo do financiamento e a Regra do 70 *(opcional)*

Não vale ponto e não entra na assinatura. É para quem terminou cedo e quer
brincar com os dois operadores que só fazem sentido com inteiros, `//` e `%`,
e com o atalho mental mais famoso da economia.

O município contratou um financiamento de infraestrutura em `MINHAS_PARCELAS`
parcelas mensais de `MINHA_PARCELA` mil reais. A diretoria quer o prazo em
"anos e meses", não em parcelas.

E a **Regra do 70**: um valor que cresce a $x$% ao ano dobra em
aproximadamente $70/x$ anos. Aplique ao CAGR que você calculou no Exercício 3.

> 🔮 Na Missão 3 (`while`) você vai calcular esse número de forma **exata**,
> simulando ano a ano. Guarde o resultado para comparar.

In [ ]:
# TODO 1: anos completos e meses restantes (// e %)
anos_completos = ...
meses_restantes = ...

# TODO 2: valor total pago ao longo do contrato, em mil R$
total_pago = ...

# TODO 3: o financiamento representa quantos % do PIB total do município?
peso_no_pib_pct = ...

print(f"Prazo: {anos_completos} anos e {meses_restantes} meses")
print(f"Total pago: R$ {total_pago:.2f} mil")
print(f"Peso no PIB municipal: {peso_no_pib_pct:.2f}%")

# TODO 4: a Regra do 70 sobre o SEU CAGR do Exercício 3.
# ⚠️ Se o seu CAGR foi negativo ou muito perto de zero, a regra não se
# aplica — explique em um comentário por que, em vez de forçar a conta.
anos_para_dobrar = ...
print(f"A esse ritmo, o PIB per capita dobra em ~{anos_para_dobrar:.1f} anos.")

registrar("bonus", "...")

---

## ✍️ Questão autoral (vale 2,0)

Agora inverta os papéis: **você** escreve o exercício.

Invente uma questão nova que use as ferramentas deste bloco (tipos,
operadores, f-strings, `if`/`elif`/`else`, `try`/`except`, `raise`) sobre os
dados do **seu** município. Depois resolva-a e, o mais importante, explique
**qual erro de compreensão ela pega**.

Uma boa questão de prova não é a mais difícil — é a que separa quem entendeu
de quem decorou. A nota aqui vai muito mais para a **justificativa** do que
para a dificuldade.

> 💡 Exemplos do tipo de erro que uma questão pode "pegar": confundir `=` com
> `==`; achar que `input()` devolve número; usar `if` independentes onde
> deveria haver `elif`; esquecer o `else` e ficar com uma variável indefinida;
> somar percentuais que deveriam ser compostos; comparar dinheiro com `==`;
> inverter a ordem de uma cadeia de faixas. **Não repita esses exemplos** —
> invente o seu.

### 📝 Enunciado da sua questão

> ...

### 🎯 Que erro de compreensão ela pega, e por quê

> ...

### 🧪 Qual seria a resposta de alguém que comete esse erro

> ...

In [ ]:
# TODO: a SUA solução para a SUA questão.
# O código precisa rodar e usar os seus dados personalizados.

...

In [ ]:
# TODO: substitua a nota abaixo — como foi inventar a questão?
registrar("questao-autoral", "...")

---

## 🐞 O obstáculo — onde você travou (parte dos 1,5)

Em algum momento desta atividade alguma coisa não saiu como você esperava: uma
célula quebrou, um `conferir()` insistiu em ficar vermelho, um número saiu
absurdo. Isso acontece com **todo mundo**, sempre — é a experiência normal de
quem programa, não sinal de fraqueza.

Conte esse momento abaixo, **com as suas palavras**. **Não cole mensagem de
erro nenhuma**: o traceback não interessa aqui, e colar um não vale ponto. O
que interessa é o raciocínio que te tirou de lá.

Descrever um obstáculo com precisão é a mesma habilidade que vai te permitir
depurar sozinho quando uma função devolver algo estranho —
e é a habilidade que você usa quando precisa **pedir ajuda de forma útil**,
para um colega ou para o professor.

> ⚠️ Uma atividade entregue sem nenhum obstáculo relatado é, ela própria, um
> sinal. Se você realmente não travou em nada, escolha o ponto que exigiu mais
> atenção e conte por quê.

### 🎯 O que você estava tentando fazer

> ...

### 🔀 O que aconteceu em vez disso

*(descreva com as suas palavras: o programa parou? saiu um número estranho? o
`conferir()` reclamou de quê?)*

> ...

### 🔍 Como você descobriu a causa

*(reler o enunciado? testar um valor menor? imprimir uma variável no meio?
comparar com a Missão 1? perguntar a alguém?)*

> ...

### 🔧 O que você mudou para destravar

> ...

### 💡 Como você vai reconhecer isso da próxima vez

> ...

---

## 📓 Diário de bordo

A célula abaixo mostra o seu próprio ritmo de trabalho: quando você passou por
cada etapa e quanto tempo levou entre elas.

O diário é **seu** e está visível para você — nada é coletado às escondidas. Na
correção, ele conta pelas **notas** que você escreveu em cada `registrar()`,
não pelos horários: uma nota específica ("travei no `elif`, tinha esquecido os
dois pontos") vale mais que dez registros vazios.

Se você tiver esquecido de trocar algum `"..."` pela sua nota, volte lá agora —
os `registrar()` podem ser rodados de novo.

In [ ]:
diario()

---

## 💭 Reflexão obrigatória (§5.1 do programa)

Em **2 a 3 frases**, qual foi a principal dificuldade desta atividade?

Seja específico. "Achei difícil" não diz nada; "demorei a entender por que a
projeção B dava quase o mesmo valor do PIB atual, até perceber que `**` é
avaliado antes de `/`" diz tudo — e mostra que você aprendeu.

> ...

---

## 🤖 Declaração de uso de IA (§5.2 do programa)

Esta atividade pertence à **Unidade 1**, onde o uso de IA generativa **não é
permitido**. Assinale abaixo, substituindo `[ ]` por `[x]`:

- `[ ]` Declaro que **não utilizei** ferramentas de IA generativa (ChatGPT,
  Claude, Gemini, Copilot ou similares) na resolução desta atividade.

- `[ ]` Declaro que **utilizei** ferramenta de IA generativa, contrariando a
  política da Unidade 1. Descrevo abaixo o que fiz e por quê.

**Se utilizou, preencha a tabela** (uma linha por consulta):

| ferramenta | o que pedi (prompt literal) | o que aceitei | o que rejeitei e por quê |
|---|---|---|---|
| ... | ... | ... | ... |

**Consultas que não são IA generativa e são permitidas** — marque as que usou:

- `[ ]` módulos de slides da disciplina
- `[ ]` notebooks das Missões 1 e 2
- `[ ]` documentação oficial do Python
- `[ ]` colegas de turma
- `[ ]` outras: ...

> 💬 A declaração honesta de uso, mesmo contrariando a política, é tratada como
> questão pedagógica — a ser conversada. A declaração falsa é outra coisa.

---

## 🔏 Assinatura e push

Rode a célula abaixo **por último**, depois de todos os `conferir()` estarem
fechados. Ela resume a sua matrícula, as respostas que passaram pela
conferência e o seu diário.

O contador `n/n conferidos` faz parte da linha. Se ele vier incompleto, algum
exercício ficou para trás — volte, ajuste e rode o `conferir()` daquela etapa
de novo antes de assinar. (O Bônus não entra na conta.)

In [ ]:
assinatura()

### 🚀 Fechando a entrega

Antes do commit final, confira as três coisas:

- `[ ]` o notebook foi rodado **inteiro, de cima para baixo**, e **salvo** com
  as saídas visíveis;
- `[ ]` todos os campos de texto estão preenchidos — previsão, questão
  autoral, obstáculo, reflexão e declaração de IA;
- `[ ]` os `registrar()` têm notas suas, não os `"..."` que vieram no template.

Agora, no terminal, dentro da pasta do repositório:

```bash
git add atividade-01-a-regua-do-observatorio.ipynb
git commit -m "ASSINATURA: <cole aqui a linha impressa acima>"
git push
```

Confira em seguida, na página do repositório no GitHub, se o notebook aparece
com as saídas. **O `git push` guarda o seu trabalho, mas não é a entrega
oficial** — falta um passo:

> 🎯 Vá ao **Google Sala de Aula**, abra a atividade "Relatório 01" e cole o
> endereço do seu repositório no GitHub. Só depois disso a entrega está
> feita.

> 🔒 O arquivo `diario-r01.json` fica de fora do commit de propósito (está no
> `.gitignore`): ele é o seu rascunho local. O que o professor lê é a tabela do
> `diario()` impressa no notebook.

> 💬 A correção volta como **nota e comentário no Google Sala de Aula** — é lá
> que a conversa continua, e é lá que você responde se discordar de algum
> ponto.